### ISSUE TITLE
End-to-end training using ultrasound image features

### ISSUE THREAD

**Issue by amadeusuzx** [espnet/espnet#1926]
Hi
We are working on a silent speech recognition project and recently made a try on ESPnet training using ultrasound image features. The dataset we used is SSC archive(Silent Speech Challenge, [https://ftp. espci.fr/pub/sigma/](url)), which contains only about  2350 recordings of sentences and corresponding ultrasound tongue imaging data. We replaced MFCC with features extracted from those ultrasound data using auto-encoder and DCT respectively, and other procedure remains the same as wsj/asr1 recipe(including the pretrained LM). 
Maybe there is something wrong in our implement, we got some terrible results.
Could you please give me some advices on this work?
![image](https://user-images.githubusercontent.com/44519021/81920520-4509bd80-9614-11ea-9a68-41e98e7a95df.png)

CER (or TER) result
```
| SPKR    | # Snt   # Wrd   | Corr      Sub     Del     Ins     Err   S.Err  |
| Sum/Avg |  231    11923   | 57.8     31.7    10.6    42.0    84.2   100.0  |
```
WER result 
```
|  SPKR    | # Snt    # Wrd  |  Corr     Sub      Del      Ins     Err    S.Err  |
|  Sum/Avg |  231      2067  |  21.1    76.5      2.3     61.8   140.6    100.0  |
```
 and some decode examples:
```
INFO: input lengths: 72
INFO: max output length: 72
INFO: min output length: 0
INFO: groundtruth: PENGUINS LIVE NEAR THE ICY ANTARCTIC.
INFO: prediction : BACK IN LIFE AND HIS HAND THERE AND INTERESTING G.<eos>
INFO: input lengths: 68
INFO: max output length: 68
INFO: min output length: 0
INFO: groundtruth: GUESS THE QUESTION FROM THE ANSWER.
INFO: prediction : THE CASE THOUGH THE FOUR FROM THE AND THE S.<eos>
```
Since SSR(silent speech recognition) datasets are usually small due to the difficulty of SSC data acquisition, do you think it is possible to train a SSR model using end-to-end methods?

**Comment by sw005320** [comment]
First, thanks for applying espnet to a very interesting project!
Frankly, the amount of training data is a bit too small.
Also, the WSJ language model is very different from your sentences I guess.
My suggestion is to start it without LM.
Then, we could specify the problem more.

**Comment by amadeusuzx** [comment]
Thanks for you advice. I'll try it and put further results here.

**Comment by amadeusuzx** [comment]
I decoded test dataset features without LM, and here are some results:
```
"rec_text": MANY DEALERS SAID IT WAS MEANT ON HIGHER EARNING IN CERTAIN O.<eos>
"text":     MEDIEVAL SOCIETY WAS BASED ON HIERARCHIES.
"score":    -228.93271662294865

"rec_text": PROSPECTIVELY AND WAS PRESENTED TO THE COLONIAL E.<eos>
"text":     PROJECT DEVELOPMENT WAS PROCEEDING TOO SLOWLY.
"score":    -157.9039665311575

"rec_text": SINGLE CALL PROGRAMS WILL BE COMPILED BY ALLEN M.<eos>
"text":     CYCLICAL PROGRAMS WILL NEVER COMPILE.
"score":    -170.19381120428443

"rec_text": CHARACTER ACTION IN THE NOVEL INSTRUCTIONS IS A ROLE REALLY SHARE.<eos>
"text":     CORRECT EXECUTION OF MY INSTRUCTIONS IS CRUCIAL.
"score":    -210.3042588122189

"rec_text": MOST PRICES AND THERE ARE FOUR CONTENTIOUS CONTENTIOUS S.<eos>
"text":     MOST PRECINCTS HAD A THIRD OF THE VOTES COUNTED.
"score":    -200.74529799818993

"rec_text": WHILE WHAT FOR IMPERTINENT IS THERE THE SCREAMED IN THE S.<eos>
"text":     WHILE WAITING FOR CHIPPER SHE CRISSCROSSED THE SQUARE MANY TIMES.
"score":    -176.09766364842653
```
Do you think there are any problems in the training stage?

**Comment by Fhrozen** [comment]
Hello there.
Thank you for the interesting project.
Just as comment, did u try to augment your data set? Not sure if it is going to work but, for images, I think you could use cropping, rotating, mirroring, adding noises (white, salt&pepper, ...) and shifts. Then, maybe you could have 4x times your dataset. 

The performance can be affected by the extension of your dataset as @sw005320 mentioned. 
Good Luck

**Comment by sw005320** [comment]
I would say that it's working given such small data.
CER 40% is not a bad start.
Could you show us the training and decoding config files?
We could give you further advice.

As @Fhrozen mentioned, data augmentation would be an interesting direction as well.

**Comment by amadeusuzx** [comment]
@sw005320 @Fhrozen Thanks for your concerns and great advices! I'll go on trying to improve the accuracy on SSC dataset.
And here are the tuning config files:
Decode yaml
```
batchsize: 0
beam-size: 10
penalty: 0.0
maxlenratio: 0.0
minlenratio: 0.0
ctc-weight: 0.3
lm-weight: 1.0
```
Training yaml
```
# network architecture
# encoder related
elayers: 12
eunits: 2048
# decoder related
dlayers: 6
dunits: 2048
# attention related
adim: 256
aheads: 4

# hybrid CTC/attention
mtlalpha: 0.3

# label smoothing
lsm-weight: 0.1
# minibatch related
batch-size: 32
maxlen-in: 512  
maxlen-out: 150 

# optimization related
opt: noam
accum-grad: 2
grad-clip: 5
patience: 3
epochs: 100
dropout-rate: 0.1

# transformer specific setting
backend: pytorch
model-module: "espnet.nets.pytorch_backend.e2e_asr_transformer:E2E"
transformer-input-layer: conv2d    
transformer-lr: 10.0
transformer-warmup-steps: 25000
transformer-attn-dropout-rate: 0.0
transformer-length-normalized-loss: false
transformer-init: pytorch

# Report CER & WER
report-cer: true
report-wer: true
```

**Comment by sw005320** [comment]
OK, these look normal. 
Then, the next action is to tune the optimizer related options by carefully monitoring the learning curve.
Suppose that the current learning curve is OK, you can try to find hyper-parameters getting better behaviors from the current one.

You could increase `batch-size` as much as possible, tune, `transformer-lr` from 1 to  10, `accum-grad` from 2 to 8, and `transformer-warmup-steps` (this depends on the data and model size).
Note that these hyperparameters depend on each other, unfortunately, but when tuning these parameters, you'd get more sense of how they depend on each other.
The nice part of the transformer is that you don't have to tune the model topology so much.

Do you think you can get some text-only data that is close to this content?
If so, we could also build an LM. Probably, we may also make some improvements by using the current transcriptions? You can just use the transcriptions as your training data in the wsj `run.sh` (probably you can just remove WSJ text data, i.e., you just exclude `${lmdatadir}/train_others.txt`  in the LM training data preparation in https://github.com/espnet/espnet/blob/master/egs/wsj/asr1/run.sh#L173-L178
Again, if you would have some text-only data matching to this task, you can add it instead of `${lmdatadir}/train_others.txt`.

**Comment by amadeusuzx** [comment]
@sw005320 Thank you very much!!
I'll try to update parameters and build a LM. Further results will be put here!

**Comment by amadeusuzx** [comment]
@sw005320
The scripts are selected from TIMIT corpus and I get the whole dataset with all 6300 scripts. Hopefully I can build a much better LM. Also thank you so much for providing advices on tuning the model. I'll try to improve it and further results will be put here!

**Comment by sw005320** [comment]
Yes, timit only has phone n-gram models, because it is designed for phoneme recognition. If your goal is phoneme recognition, you may also use it, but it would not be your goal I guess.
As you may know, timit is based on a phonetically balanced sentence, and probably it would not be easy to find external text data that matches such specific situations. Let's try to build an LM with the 6300 transcriptions.

**Comment by amadeusuzx** [comment]
Sorry for the late reply!
As a result, it seems hard to improve the model further by tuning.
I tried: 

- **Increase` batch-size` from 32 to 64 and 128.**

Both resulted in training stopping earlier. Instead,  batch size 16 turn out to be the best option, where the accuracy reached 55% in training step.
![image](https://user-images.githubusercontent.com/44519021/82992475-31118300-a03a-11ea-9cc3-040c1d21c0d0.png)

```
(Decode using timit corpus LMs )
    | SPKR    | # Snt   # Wrd   | Corr     Sub     Del     Ins     Err   S.Err  |
 CER| bruce   |  231    11923   | 63.0    22.1    14.9    10.2    47.1   100.0  |
 WER| bruce   |  231    2067    | 33.8    61.0    5.2     17.9    84.2   100.0  |
```

- **Change `accum-grad` to 4,6,8**
This made training stop with in first 8 epochs.

- **`transformer-warmup-steps`**
I couldn't figure out what this parameter has influenced training...


Besides, I also tried decode the training data and found that the result was much better(maybe it's just a matter of course but) I wonder if there is anyway to use espnet to decode train data so that we can get aligned texts corresponding to the input? Because in further work we plan to use Lipnet, which requires aligned data.
Thanks for your help again!

**Comment by sw005320** [comment]
I feel that the learning curve is not still ideal. learning rate decay seems to happen around 5 epochs but it should be happened later or may need to have a larger learning rate before the decay. Could you increase it (you can find it in https://github.com/espnet/espnet/blob/master/egs/voxforge/asr1/conf/tuning/train_pytorch_transformer_d6-2048.yaml#L34)

The voxforge Italian recipe is tuned with a small amount of training data (it has more than 10 hours though). So, you may refer to this config...
Also, the voxforge Italian recipe requires more than 100 epochs (200 epochs are enough to get a reasonable number).

**Comment by amadeusuzx** [comment]
Voxforge recipe seems great! Thank you!

Besides, we plan to try Lipnet in future work and it requires aligned data. Could you please tell me if it is possible to use ESPnet to obtain the time stamp of each character or word so that we can align text and input data?

**Comment by sw005320** [comment]
This is not straightforward, but if you check the CTC output (e.g., with https://github.com/espnet/espnet/blob/6f6452a3891d11b0b0306084dd46a464ddf8e25c/espnet/nets/pytorch_backend/e2e_asr_transformer.py#L261), you can get a frame-by-frame label output; thus you can get alignment information. Note that the CTC alignment is not very synchronized, but better than nothing.

Gettig alignment information from the attention weight is noisier, and I recommend you to try CTC one.

**Comment by amadeusuzx** [comment]
I tried Voxforge recipe and this is the best rusults, with batchsize 12, transformer lr 20.
There is another change that I used all data where transcripts are from TIMIT corpus, and tested the result using data where transcripts are from WSJ corpus (100 utterances).So I get back to WSJ pretrained LM again.


```
       -----------------------------------------------------------------.
      |                          cer_result                             |
      |-----------------------------------------------------------------|
      | SPKR   | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
      |--------+--------------+-----------------------------------------|
      | bruce  |  100    5767 | 81.3    8.1   10.5    4.0   22.6   90.0 |
      |=================================================================|
      | Sum/Avg|  100    5767 | 81.3    8.1   10.5    4.0   22.6   90.0 |
      |=================================================================|
      |  Mean  |100.0  5767.0 | 81.3    8.1   10.5    4.0   22.6   90.0 |
      |  S.D.  |  0.0    0.0  |  0.0    0.0    0.0    0.0    0.0    0.0 |
      | Median |100.0  5767.0 | 81.3    8.1   10.5    4.0   22.6   90.0 |
      `-----------------------------------------------------------------'
      ,-----------------------------------------------------------------.
      |                          ./wer_result                           |
      |-----------------------------------------------------------------|
      | SPKR   | # Snt  # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
      |--------+--------------+-----------------------------------------|
      | bruce  |  100    1023 | 63.1   30.4    6.5    5.2   42.0   90.0 |
      |=================================================================|
      | Sum/Avg|  100    1023 | 63.1   30.4    6.5    5.2   42.0   90.0 |
      |=================================================================|
      |  Mean  |100.0  1023.0 | 63.1   30.4    6.5    5.2   42.0   90.0 |
      |  S.D.  |  0.0    0.0  |  0.0    0.0    0.0    0.0    0.0    0.0 |
      | Median |100.0  1023.0 | 63.1   30.4    6.5    5.2   42.0   90.0 |
      `-----------------------------------------------------------------'
```
![image](https://user-images.githubusercontent.com/44519021/83367379-36cae800-a3ef-11ea-8558-5aba634d6c58.png)
Accuracy on train set bacame over 95%, and character-level accuracy is quite good considering such small dataset. for such small dataset. But I think WER is still not satisfying though. Can you give me some advice again?

**Comment by kan-bayashi** [comment]
From your training curve, the model is not converged yet.
Why don't you continue to train until, say, 2 or 300 epochs?
And did you try with SpecAug? 
I'm not sure it is working well with your image-based features, but I think it is worthwhile to try.

**Comment by amadeusuzx** [comment]
@kan-bayashi 
Thanks for telling me about SpecAug! I never knew it ..
About the training curve, I set it to 200 before and found it really stopped after about 100 epochs, then changed it to 100.
Actually the curve is like
![image](https://user-images.githubusercontent.com/44519021/83405901-8093ec80-a448-11ea-8d9b-464cf24c5d3b.png)
after 100 epochs..
Could you please tell me if word accuracy 63.1% is normal when we have character accuracy of 81.3%?
BTW, here are some decode examples.
```
Speaker sentences   0:  bruce   #utts: 100
id: (bruce-bruce_list01_test01)
Scores: (#C #S #D #I) 4 5 3 0
REF:  AT N. E.  C.     the need for INTERNATIONAL MANAGERS                will KEEP RISING  
HYP:  ** ** AND INDEED the need for ************* INTERNASIONALOPETAGUERS will A    PRICING 
Eval: D  D  S   S                   D             S                            S    S       

id: (bruce-bruce_list01_test02)
Scores: (#C #S #D #I) 7 1 1 0
REF:  about half these managers are in the U. S.   
HYP:  about half these managers are in the ** WAYS 
Eval:                                      D  S    

id: (bruce-bruce_list01_test03)
Scores: (#C #S #D #I) 7 0 0 0
REF:  the company previously traded over the counter 
HYP:  the company previously traded over the counter 
Eval:                                                

id: (bruce-bruce_list01_test04)
Scores: (#C #S #D #I) 10 0 0 0
REF:  two other issues began trading recently on the big board 
HYP:  two other issues began trading recently on the big board 
Eval:                                                          

id: (bruce-bruce_list01_test05)
Scores: (#C #S #D #I) 3 1 0 1
REF:  *** UNION officials expect ratification 
HYP:  YOU NUT   officials expect ratification 
Eval: I   S                                   

id: (bruce-bruce_list01_test06)
Scores: (#C #S #D #I) 1 7 0 1
REF:  a **** P.   B.   G. C.   SPOKESWOMAN DECLINED COMMENT 
HYP:  a MERE MERE GEIS IS MOST WOMEN       DECLINE  COMPANY 
Eval:   I    S    S    S  S    S           S        S       

id: (bruce-bruce_list01_test07)
Scores: (#C #S #D #I) 6 2 1 3
REF:  ** * ANALYSTS too ** generally played down the EFFECT ON       banks 
HYP:  AT A LEAST    too AS generally played down the ****** OFFICTON banks 
Eval: I  I S            I                            D      S              

id: (bruce-bruce_list01_test08)
Scores: (#C #S #D #I) 7 3 0 0
REF:  the INVESTOR  now OWNS SEVENTY three percent of the company 
HYP:  the INVESTORS now ONES FIND    three percent of the company 
Eval:     S             S    S                                    

id: (bruce-bruce_list01_test09)
Scores: (#C #S #D #I) 4 3 1 1
REF:  texaco has three ** CHOICES A  company ADVISER SAYS     
HYP:  texaco has three TO IS      IN company ******* ADVISERS 
Eval:                  I  S       S          D       S        

id: (bruce-bruce_list01_test10)
Scores: (#C #S #D #I) 12 0 0 0
REF:  the rise in auto imports also reflects higher prices for imported cars 
HYP:  the rise in auto imports also reflects higher prices for imported cars 
Eval:                                                                        

id: (bruce-bruce_list01_test11)
Scores: (#C #S #D #I) 12 2 0 0
REF:  WHAT we don't know is how much is price and how much is VOLUME 
HYP:  ONE  we don't know is how much is price and how much is FLUME  
Eval: S                                                       S      
```

I will now try SpecAug for data augmentation!

**Comment by sw005320** [comment]
First, it's better to use the error rate than accuracy to consider insertion, deletion, and substitution errors.
I think your numbers are quite reasonable by considering a very small amount of training data. 
22.6 CER <-> 42.0 WER is a reasonable.

Also, as you may notice, WSJ has a lot of abbreviations (e.g., NEC, PBGC, and US) and this was converted with the period based representation (i.e., NEC -> N. E. C.). If your TIMIT training data does not include such phrases, e2e model cannot do anything, and this directly becomes the error. This seems to happen a lot in your cases. Without this abbreviation issue, your model is almost correct.

We could still have more efforts by specaug, the above text normalization, and LM but I feel that your e2e model is getting quite reasonable numbers already.

Is your data public?
Is so, is it possible to make a PR?
We're very interested in this setup, result, and model.
We can also give you further comments directly to your source codes via PR.

**Comment by amadeusuzx** [comment]
The abbreviations do hinder the model's performance. but it seems not a big matter because the number of abbreviations is quite small in my test set. (only 22 abbreviation characters included). Aslo I removed abbreviation sentences then calculated WER again:
```
| SPKR   | # Snt # Wrd | Corr    Sub    Del    Ins    Err  S.Err |
| bruce  |   91    907    | 64.4   29.8    5.8    5.5   41.1   89.0 |
```
Thanks for your invitation for a PR! However I have to do it tomorrow for some other work. Looking forward to your comments!

**Comment by supernaiter** [comment]
Hi, nice to meet you! @sw005320

**Comment by supernaiter** [comment]
Hi nice to meet you @sw005320  !
I'm a co-worker of @amadeusuzx , and writing a brief paper about this project. I really appreciate your continuous support!

I have a question, is CER calculated with the output of lm decoding?

We had thought CER is calculated using raw outputs of E2E model without lm decoding, but we have tested decoding with some different language models and got different results on CER.

Thanks!

**Comment by sw005320** [comment]
Hi @supernaiter,

CER was computed with LM decoding.
The recognition results are the same for the WER and CER computations, and we just used different units to compute them.
